drivers/console/ipm_console.c

```diff
@@ -52,22 +53,30 @@ static void msgq_thread(void *arg1, void *arg2, void *arg3)
 
 static int console_out(int c)
 {
+	static char buf[CONFIG_IPM_CONSOLE_LINE_BUF_LEN];
+	static size_t len;
 	int ret;
 
+	if (c != '\n' && len < sizeof(buf)) {
+		buf[len++] = c;
+		return c;
+	}
+
 #if defined(CONFIG_IPM_CONSOLE_IMMEDIATE)
-	ret = ipm_send(ipm_dev, 1, c & ~BIT(31), NULL, 0);
+	ret = ipm_send(ipm_dev, 1, len & ~BIT(31), buf, len);
```

Comment:
I'm seeing this pattern a lot in this pull request, where you're setting the high bit of the IPM message ID.   Now, that's fine as far as IPM itself goes -- the ID is just an opaque value.  But clearly it means something, and I can't see what.

What's going on here?  This needs documentation, I think.  Right now the code you're adding is speaking an "ADSP IPM" variant protocol and won't work with default IPM backends.  That's not crazy, since they're all device-specific, but it needs some kind of framework in software I think and not just a bunch of raw bits flying around.

Reply:
Here I have changed, following your suggestion, from sending one character to sending a line. This in not IPM but user-defined "console over IPM", kind of higher level protocol.

Reply:
OK, then maybe a "IPM_CONSOLE_MSGID(len)" macro would make this clearer?

Reply:
Oh, and note that this presumes a 32-bit-clean IPM message ID.  Note that IPM implementations are allowed to have a smaller ID space, there's an API to query it.  This should probably be matched with an __ASSERT (or BUILD_ASSERT) to make sure no one tries this with an incompatibile backend.

Reply:
OK, I've figured out what you're doing.  You're clearing this bit because the hardware reserves the top bit of the message register for the BUSY flag and you don't want to clobber anything.  But the API already gives you a way to inform users of that: `ipm_max_id_val_get()`.  So implement that API and put this bit clear down into the driver level underneath `ipm_send()`.